Adds content to page component react examples. (#8198)
patternfly/patternfly-react: packages/react-core/src/components/Page/examples/Page.md

## Examples

### Basic page components

A page will typically contain the following components: 

> - A `<Page>` with a `header` that often contains a masthead, a `<PageSidebar>`, and one or more `<PageSection>` components. 

Reviewer: I also wonder if this should be updated to be a list?

<img width="860" alt="Screenshot 2022-11-18 at 11 09 45 AM" src="https://user-images.githubusercontent.com/35148959/202762250-1400f841-327a-4d9a-ba41-d3d3ec88318d.png">

> -[Mastheads](/components/masthead) contain the `<PageToggleButton>`, a `<MastheadMain>` that contains a `<MastheadBrand>`, and the page's header toolbar within `<MastheadContent>`. 

Reviewer: I left a comment about making this a list, but this also looks like a nested bullet so if we keep it like this, I think it would be 

```suggestion
  - [Mastheads](/components/masthead) contain the `<PageToggleButton>`, a `<MastheadMain>` that contains a `<MastheadBrand>`, and the page's header toolbar within `<MastheadContent>`. 
```

### Vertical navigation

To add vertical side navigation to a `<Page>`, pass a `<PageSidebar>` component into the `sidebar` property. To render navigation in the `<PageSidebar>`, use the `nav` property. 

> The `isNavOpen` property helps facilitate the opening and closing of the sidebar and should be 'true' when the navigation sidebar is visible. 

Reviewer: I could be off base here but I don't think `<PageSidebar>` currently allows for anything other than including whatever is passed to the `nav` prop. Though it doesn't look like that has to be `<Nav>`

https://github.com/patternfly/patternfly-react/blob/e1f95ce96852165dd4d102cd87c6b49288f01eff/packages/react-core/src/components/Page/PageSidebar.tsx#L44-L61

> The `isNavOpen` property helps facilitate the opening and closing of the sidebar and should be 'true' when the navigation sidebar is visible. 

Reviewer 2: Yeah it would be worth aligning on, especially whether it should be `children` or remain `nav`.

> The `isNavOpen` property helps facilitate the opening and closing of the sidebar and should be 'true' when the navigation sidebar is visible. 

Reviewer 3: So technically you can pass any React node to the `nav` prop.  I suppose by using the `nav` it is more explicit to the consumer that the a nav is what is expected.  @mcoker would part of the intention of renaming the `nav` prop to `children` be to allow more explicit flexibility?

> The `isNavOpen` property helps facilitate the opening and closing of the sidebar and should be 'true' when the navigation sidebar is visible. 

Reviewer: @tlabaj yep, I think so. @mattnolting is currently working on a PR to allow for multiple sidebar body elements and sidebar body modifier classes, so we could hold off on this until that work is complete. https://github.com/patternfly/patternfly/pull/5093/files

```ts file="./PageVerticalNav.tsx"
```

> ### Legacy vertical navigation

Reviewer: As far as I can tell, this is just an example of using `<PageHeader>` instead of `<Masthead>`, which shouldn't impact vertical navigation

```suggestion
### Legacy page header
```

This example shows the legacy implementation of a page's vertical navigation. Our updated recommendation advises you to use a masthead and toolbar to make headers, rather than `<PageHeader>` and `<PageHeaderTools>` as shown in the following example.

```ts file="./PageVerticalNavUsingPageHeaderComponent.tsx"
```

### Horizontal navigation

> To add horizontal top navigation to a `<Page>`, each navigation item should be set as a `<ToolbarItem>` in the `<Toolbar>` that is passed to the `<Masthead>`.

Reviewer: I don't know that this makes sense. The whole nav is a toolbar item, not each nav item. The structure is like:

```
masthead
  toolbar
    toolbar-item
      nav
        nav-list
          nav-item
          nav-item
          nav-item
          ...
```

Though I don't think we need to go into this detail here unless this is specific to the page component. From the perspective of the page component, horizontal nav is an absence of the sidebar and hamburger/toggle button.

```ts file="./PageHorizontalNav.tsx"
```

> ### Tertiary navigation

Reviewer: The tertiary nav is a legacy variant of navigation. It still exists, but we don't recommend anyone use it - we recommend horizontal subnav instead. Also this and horizontal subnav should be paired with either vertical and/or horizontal nav, otherwise you'd just make whatever is in tertiary nav the main vertical/horizontal nav, so I'd say it's probably good to update this example to include the pagesidebar to indicate there is also vertical nav?

> To add a tertiary navigation component to a page, use the `tertiaryNav` property. The following example passes a string of ‘Navigation’ to `tertiaryNav`, but you can pass fully built-out components, such as navigation.

Reviewer: Passing navigation is what this is for - the reason this is a unique page section is because it provides spacing that works with the tertiary nav in context of other stuff - breadcrumbs, tabs, other page sections, the tertiary nav horizontal scroll buttons...

Technically you could put something else in here if you wanted, but that's probably not advised in most instances. I think we should communicate that it's for tertiary nav. Inline with my comment above, we have a unique page section for breadcrumbs, tertiary nav, horizontal nav, tabs... they're ideally designed for those respective components.

> In this example, `<Masthead>` sets the `display` property equal to `"{{ default: 'stack' }}"` to stack the masthead items and also passes in `"inset={{ default: 'insetXs' }}"` to adjust the inset padding.

Reviewer: I think you can remove this. I don't see it, and the details of masthead shouldn't matter - tertiary nav is rendered in the page content, not the header/masthead.

```ts file="./PageTertiaryNav.tsx"
```

> ### Uncontrolled nav

Reviewer: seems like we're `s/nav/navigation/`'ing in other places?

```suggestion
### Uncontrolled navigation
```

When the `isManagedSidebar` property is true, it manages the sidebar open/close state, removing the need to pass both `isNavOpen` into the `<PageSidebar>` and `onNavToggle` into the `<PageHeader>`.

```ts file="./PageUncontrolledNav.tsx"
```

### Filled page sections

To make a `<PageSection>` fill the available vertical space of the page, set the `isFilled` property equal to true.

> By default, the last page section sets `isFilled` equal to true.

Reviewer: This isn't entirely accurate... there is the concept of a "filled" page section, meaning it fills the available space, and an `isFilled` boolean prop that applies to page sections to change the default "fill" behavior of any page section.

In the core CSS (without any consideration to the react `isFilled` prop), only the last page section is "filled" by default, but that's done via CSS and not via the `isFilled` prop. So the last section is "filled" but it's not `isFilled`.

What `isFilled` allows a user to do is to change the default "filled" behavior from CSS - so they could, for example, make their first section "fill" by setting `isFilled` and disable the last section from being filled by default by setting `isFilled={false}`.

```ts file="./PageWithOrWithoutFill.tsx"
```

### Main section padding

To adjust the padding of a `<PageSection>`, you can pass in different values to the `padding` property. These value should be aligned to various breakpoints, including 'default', 'sm', 'md', 'lg', 'xl', '2xl'. 

To remove padding, pass 'noPadding' into `padding`. For example, the second section in this example passes in `padding={{ default: 'noPadding' }}`. Since no specific breakpoints are mentioned, every breakpoint will have 'noPadding'.

> To add padding on specific breakpoints, specify the value for those breakpoints. For example, the third section in this example passes in `padding={{ default: 'noPadding', md: 'padding', lg: 'padding' }}`. On 'md' and 'lg' breakpoints, the default value will be overwritten, and padding will be added.

Reviewer: The way this is worded, and the way the modifiers are used currently seems to imply that if you set padding/nopadding at a breakpoint, it applies to _that breakpoint only_. FWIW, as the breakpoints increase (the page viewport width increases), each breakpoint's padding value inherits from the previous breakpoint. In this example, you'll notice that whatever padding is set on `lg` also applies to `xl` and `2xl`. Which means that `md - padding, lg - padding` is redundant - you only need `md - padding` in that example, because `lg` and above will inherit from `md`.
 
The page section in the example that shows this has the text "Section with padding only on medium/large," so I think a good way to show this would be `default - noPadding, md - padding, xl - noPadding`. Though there is a blurb below about removing padding at a breakpoint, so my suggestion here could be out of order with that.

To remove padding at specific breakpoints, pass in 'noPadding' at those breakpoints. For example, the fourth section in this example passes in `padding={{ md: 'noPadding' }}`, which means that 'md' is the only breakpoint that will remove padding.

```ts file="./PageMainSectionPadding.tsx"
```

### Group section

To group page content sections, add one or more `<PageGroup>` components to a `<Page>`. 

> The following example adds a group containing both the `<PageNavigation>` and a section of content.

Reviewer: FWIW this group has `<PageNavigation>`, `<PageBreadcrumb>` and `<PageSection>`. It's also worth noting that you can group things in a `<PageGroup>` via `isBreadcrumbGrouped`, `isTertiaryNavGrouped`, and `additionalGroupedContent `. Not sure if that's relevant for this example, but wanted to mention it.

```ts file="./PageGroupSection.tsx"
```

### Centered section

When a width-limited page section is wider than the value of --pf-c-page--section--m-limit-width--MaxWidth, the section will be centered in the main section.

> The content in this example is placed in a card to better illustrate how the section behaves when it is centered. A card is not required to center a page section.

Reviewer 2: I'm actually not seeing any difference in this example between having `isWidthLimited` and `isCenterAligned` applied vs removing them. I think a follow up to make the purpose of the props used more clear would be good, unless I'm missing the intent.

We should probably avoid using the same text from the example, though, and just mention what props to set to center the PageSection

Reviewer 4: @mcoker WDYT?

Reviewer: Here's the difference:

Default:
<img width="2609" alt="Screenshot 2022-11-18 at 11 56 21 AM" src="https://user-images.githubusercontent.com/35148959/202771355-a1280473-eb6a-4257-babe-cef844d62609.png">

`isWidthLimited`:
<img width="2609" alt="Screenshot 2022-11-18 at 11 56 42 AM" src="https://user-images.githubusercontent.com/35148959/202771382-2d6c4f1f-7864-445d-bdf1-92d9816fdb7a.png">

`isCenterAligned`:
<img width="2609" alt="Screenshot 2022-11-18 at 11 56 49 AM" src="https://user-images.githubusercontent.com/35148959/202771423-51395e61-c291-407e-b3a2-c550bc8a11c6.png">

Core has an example that shows a center aligned section with (first section) and without (second, card gallery) a card wrapping the content to shows the boundary. We also have a third section that I don't know is needed, but shows how you can center the text, though it uses a utility class that isn't tied to the page component CSS at all.

https://www.patternfly.org/v4/components/page/html-demos/centered-section/

To add vertical side navigation to a `<Page>`, pass a `<PageSidebar>` component into the `sidebar` property. To render navigation in the `<PageSidebar>`, use the `nav` property. 

The `isNavOpen` property helps facilitate the opening and closing of the sidebar and should be 'true' when the navigation sidebar is visible. 

```ts file="./PageVerticalNav.tsx"
```

### Legacy vertical navigation

This example shows the legacy implementation of a page's vertical navigation. Our updated recommendation advises you to use a masthead and toolbar to make headers, rather than `<PageHeader>` and `<PageHeaderTools>` as shown in the following example.

```ts file="./PageVerticalNavUsingPageHeaderComponent.tsx"
```

### Horizontal navigation

To add horizontal top navigation to a `<Page>`, each navigation item should be set as a `<ToolbarItem>` in the `<Toolbar>` that is passed to the `<Masthead>`.

```ts file="./PageHorizontalNav.tsx"
```

### Tertiary navigation

To add a tertiary navigation component to a page, use the `tertiaryNav` property. The following example passes a string of ‘Navigation’ to `tertiaryNav`, but you can pass fully built-out components, such as navigation.

> In this example, `<Masthead>` sets the `display` property equal to "{{ default: 'stack' }}" to stack the masthead items and also passes in "inset={{ default: 'insetXs' }}" to adjust the inset padding.

Reviewer 4: This line needs to be updated to:

In this example, `<Masthead>` sets the `display` property equal to `"{{ default: 'stack' }}"` to stack the masthead items and also passes in `"inset={{ default: 'insetXs' }}"` to adjust the inset padding.

(so that there are tics around the double quotes). That will fix the build issue.

## Examples

### Basic page components

A page will typically contain the following components: 

- A `<Page>` with a `header` that often contains a [masthead](/components/masthead)
  - Mastheads contain the `<PageToggleButton>`, a `<MastheadMain>` that contains a `<MastheadBrand>`, and the page's header toolbar within `<MastheadContent>`. 
- A `<PageSidebar>` for vertical navigation 

> - One or more `<PageSection>` components

Reviewer 5: PatternFly uses numerals, so even though 1 looks funny, I'd change "one" to "1" for consistency

### Vertical navigation

To add a vertical sidebar to a `<Page>`, pass a `<PageSidebar>` component into the `sidebar` property. To render navigation in the sidebar, use the `nav` property of the `<PageSidebar>`. 

The `isNavOpen` property helps facilitate the opening and closing of the sidebar and should be 'true' when the navigation sidebar is visible. 

```ts file="./PageVerticalNav.tsx"
```

### Legacy page header

This example shows the legacy implementation of a page's vertical navigation. Our updated recommendation advises you to use a masthead and toolbar to make headers, rather than `<PageHeader>` and `<PageHeaderTools>` as shown in the following example.

```ts file="./PageVerticalNavUsingPageHeaderComponent.tsx"
```

### Horizontal navigation

To add horizontal top navigation to a `<Page>`, each navigation item should be set as a `<ToolbarItem>` in the `<Toolbar>` that is passed to the `<MastheadContent>` of the `<Masthead>`.

```ts file="./PageHorizontalNav.tsx"
```

### Legacy tertiary navigation

> Horizontal sub-navigation is now recommended instead of tertiary-level navigation. 

Reviewer: Looks like we don't have a horizontal subnav example - should we have one?

Author: I could link to https://www.patternfly.org/v4/components/navigation#horizontal-subnav?

Reviewer: @edonehoo sounds good to me

> Tertiary navigation allows you to add an additional navigation menu alongside vertical or horizontal navigation. To create tertiary navigation, use the `tertiaryNav` property. The following example passes `tertiaryNav="Navigation"` into a `<Page>` component.

Reviewer: FWIW it can also be created via

```
<PageNavigation>[nav]</PageNavigation>
<PageSection type="nav">[nav]</PageSection>
```

Author: Good to know! Is this a common implementation? Generally, I haven't mentioned alternative implementation options unless they're present within the example. So (unless I'm missing this in another example) I wonder if this is worth adding a whole new example?

Reviewer: I think it's fine for now. cc @tlabaj in case she'd like to add that. I can add it in https://github.com/patternfly/patternfly-react/pull/8485 if we want to add it.

Reviewer 3: I do not think it is worth adding another example since tertiary nav is not recommended.

```ts file="./PageTertiaryNav.tsx"
```

### Uncontrolled navigation

When the `isManagedSidebar` property is true, it manages the sidebar open/close state, removing the need to pass both `isNavOpen` into the `<PageSidebar>` and `onNavToggle` into the `<PageHeader>`.

```ts file="./PageUncontrolledNav.tsx"
```

### Filled page sections

> By default, the last page section is "filled", meaning it fills the available vertifcal space of a page.

Reviewer: ```suggestion
By default, the last page section is "filled", meaning it fills the available vertical space of a page.
```

To change the default behavior, use the `isFilled` property. To make other page sections "filled", set `isFilled` equal to "true". To disable the last page section from being "filled", set `isFilled` equal to "false".

```ts file="./PageWithOrWithoutFill.tsx"
```

### Main section padding

To adjust the padding of a `<PageSection>`, you can pass in different values to the `padding` property. These values should be aligned to a specific breakpoint: 'default', 'sm', 'md', 'lg', 'xl', and '2xl'. Each breakpoint passed into the property should be given a value of either ‘padding’ or ‘noPadding’. 

As the page's viewport width increases, breakpoints inherit the padding behavior of the previous breakpoint. For example, padding that is set on 'lg' also applies to 'xl' and '2xl'. 

To remove padding entirely, pass 'noPadding' to the `default` breakpoint. For example, the second section in this example passes in `padding={{ default: 'noPadding' }}`. Since no specific breakpoints are mentioned, every breakpoint will have 'noPadding'.

> To add padding at specific breakpoints, pass in "padding" at those breakpoints. For example, the third section in this example passes in `padding={{ default: 'noPadding', md: 'padding', lg: 'padding' }}`. At 'md', 'lg',  'xl' and '2xl' breakpoints, the default value will be overwritten, and padding will be added.

Reviewer: ```suggestion
To add padding at specific breakpoints, pass in "padding" at those breakpoints. For example, the third section in this example passes in `padding={{ default: 'noPadding', md: 'padding' }}`. At 'md', 'lg',  'xl', and '2xl' breakpoints, the default value will be overwritten, and padding will be added.
```

Author: Here,  I was pulling `padding={{ default: 'noPadding', md: 'padding', lg: 'padding' }}` directly from the example. I can update my content, but do you think it's worth having someone update the example to remove `lg: 'padding'` then?

Reviewer: @edonehoo yep! I submitted a PR to remove that - https://github.com/patternfly/patternfly-react/pull/8485

> To remove padding at specific breakpoints, pass in 'noPadding' at those breakpoints. For example, the fourth section in this example passes in `padding={{ md: 'noPadding' }}`, which means that 'md', ‘lg’  'xl' and '2xl' breakpoints will have ‘noPadding’.

Reviewer: ```suggestion
To remove padding at specific breakpoints, pass in 'noPadding' at those breakpoints. For example, the fourth section in this example passes in `padding={{ md: 'noPadding' }}`, which means that 'md', 'lg', 'xl', and '2xl' breakpoints will have 'noPadding'.
```

```ts file="./PageMainSectionPadding.tsx"
```

### Group section

> To group page content sections, add one or more `<PageGroup>` components to a `<Page>`. 

Reviewer 5: same here

The following example adds a group containing `<PageNavigation>`, `<PageBreadcrumb>`, and `<PageSection>` components.

To add additional components and information to a group, you may use the following properties: 

- To indicate that a breadcrumb should be in a group, use `isBreadcrumbGrouped`. 
- To indicate that tertiary navigation should be in a group, use `isTertiaryNavGrouped`.
- To specify additional group content, use `additionalGroupedContent`. 

```ts file="./PageGroupSection.tsx"
```

### Centered section

> By default, a page section spans the width of the page. To reduce the width of a section, use the `isWidthLimited` property. To center align width-limited page sections, use the `isCenterAligned` property. When a width-limited page section is wider than the value of `--pf-c-page--section--m-limit-width--MaxWidth`, the section will automatically be centered.

Reviewer: > When a width-limited page section is wider than the value of `--pf-c-page--section--m-limit-width--MaxWidth`

I think folks might get what that's say, but it's a bit confusing because the section will never grow wider than that value. It's more true that when the page's main content area is wider than that value, the section stops growing in width as the main content area continues to grow. And if `isCenterAligned` is applied, that section will also be centered instead of left-aligned.

```suggestion
By default, a page section spans the width of the page. To reduce the width of a section, use the `isWidthLimited` property. To center align width-limited page sections, use the `isCenterAligned` property. When the main content area of a page is wider than the value of a centered, width-limited page section's `--pf-c-page--section--m-limit-width--MaxWidth` custom property, the section will automatically be centered.
```
